### DISCUSSION TITLE
Install Rush on Mac

### DISCUSSION THREAD

**Discussion by HritikR** [shreyashsaitwal/rush-cli#27] [Q&A] [answered]
How to install rush on mac os?

**Comment by shreyashsaitwal** [comment]
You can install Rush on macOS using your terminal. Read [this](https://github.com/shreyashsaitwal/rush-cli/wiki/Installation#using-bash).

**Reply by HritikR** [reply]
`hritikr@Mac Rush % curl https://raw.githubusercontent.com/shreyashsaitwal/rush-cli/main/scripts/install/install.sh -fsSL | bash
######################################################################## 100.0%##O=#  #                                                        ######################################################################## 100.0%#-#O=#  #                                                       ######################################################################## 100.0%
bash: line 42: [: /Users/hritikr/Library/Application: binary operator expected
mkdir: /Users/hritikr/Library/Application: File exists
mkdir: Support: No such file or directory`

I'm getting this error.

**Reply by shreyashsaitwal** [reply]
Try running the following before running the install script:
```.sh
mkdir -p "$HOME/Library/Application Support/rush"
```

**Reply by HritikR** [reply]
Still getting the same error

**Reply by shreyashsaitwal** [reply]
Hi @HritikR,
Can you try running this:
```.sh
curl -fsSL https://raw.githubusercontent.com/shreyashsaitwal/rush-cli/dev/scripts/install/install.sh | sh
```

**Reply by HritikR** [reply]
Installed Successfully but now getting this error when i try to run rush  `zsh: exec format error: rush`

**Reply by shreyashsaitwal** [reply]
I've pushed a change that might fix this. Can you retry running this:
```sh
curl -fsSL https://raw.githubusercontent.com/shreyashsaitwal/rush-cli/dev/scripts/install/install.sh | sh
```

**Reply by HritikR** [reply]
Still getting same error !  `zsh: exec format error: rush`

**Reply by shreyashsaitwal** [reply]
Can you paste the output of following here:
```sh
echo $(uname -sm)
```

**Reply by HritikR** [reply]
its `Darwin x86_64`

**Reply by HritikR** [reply]
@shreyashsaitwal  Any updates?

**Reply by shreyashsaitwal** [reply]
I'm not really sure why's that happening. In theory, it should download the correct executable for your architecture. Are you sure you let the download complete the last time? If not, try running the script again and wait till you get the success message. 

Ah, and sorry for the delay. I was running a bit busy lately.

**Reply by HritikR** [reply]
@shreyashsaitwal At first I tried installing using the main branch install script, that didn't work! later I tried the dev branch and it successfully got installed. I set the path using the export command and tried running rush.
Screenshot : [Attachment](https://i.imgur.com/K4IV3Dz.png)

**Reply by shreyashsaitwal** [reply]
Can you send me the Rush binary? You can find it in `$HOME/.rush/bin`.

**Reply by HritikR** [reply]
@shreyashsaitwal is it possible to  use rush with jdk version above 8?

**Reply by HritikR** [reply]
@shreyashsaitwal Rush Binary -  [Download]( https://drive.google.com/file/d/1NJfpU4uu688XovMow97b19Kw8WF8xlJ6/view?usp=sharing)

**Reply by shreyashsaitwal** [reply]
Technically, Rush doesn't actually care about what version of JDK you have, but it can't be guaranteed whether your extension will then always work in App Inventor. Also, for features like desugaring of Java 8 language features, you MUST have JDK 8, otherwise Rush will complain.

**Reply by shreyashsaitwal** [reply]
The binary you shared seems alright. This is really strange, no other Mac user has reported this issue earlier. This time, can you manually install Rush. Go through the following steps to do that:
* Delete everything inside of `$HOME/.rush` and `$HOME/Library/Application Support/rush` directories.
* Download [`rush-mac.zip`](https://github.com/shreyashsaitwal/rush-cli/releases/download/v1.2.0/rush-mac.zip).
* Extract it.
* Copy the Rush binary in `exe/mac` and place it at `$HOME/.rush/bin`
* Then, copy `dev-deps` and `tools` from the extracted `rush-mac.zip` into `$HOME/Library/Application Support/rush`.
* Done, now try to run `rush -v`.

**Reply by HritikR** [reply]
@shreyashsaitwal still same error

**Reply by shreyashsaitwal** [reply]
Then the only solution left is to build Rush from the sources. Go through the following steps to do it:
* [Install Dart](https://dart.dev/get-dart).
* Clone this repo (rush-cli).
* `cd` into it and run `chmod +x ./scripts/build.sh`
* Run `dart pub get`.
* Then, run `./scripts/build.sh -t "" -v 1.2.0`
* After the build finishes, you will find the Rush binary in `./build/bin` directory.
* Copy it to the `$HOME/.rush/bin` directory.
* Done.

**Reply by HritikR** [reply]
Error While Building -
hritikr@Mac rush-cli % ./scripts/build.sh -t "" -v 1.2.0
Generated lib/version.dart
Generated lib/env.dart
Info: Compiling with sound null safety
Error: Couldn't resolve the package 'args' in 'package:args/command_runner.dart'.
Error: Couldn't resolve the package 'dart_console' in 'package:dart_console/dart_console.dart'.
Error: Couldn't resolve the package 'rush_cli' in 'package:rush_cli/commands/build_command/build_command.dart'.
Error: Couldn't resolve the package 'rush_cli' in 'package:rush_cli/commands/create_command/create_command.dart'.
Error: Couldn't resolve the package 'rush_cli' in 'package:rush_cli/commands/migrate_command/migrate_command.dart'.
Error: Couldn't resolve the package 'rush_cli' in 'package:rush_cli/commands/upgrade_command/upgrade_command.dart'.
Error: Couldn't resolve the package 'rush_cli' in 'package:rush_cli/helpers/app_data_dir.dart'.
Error: Couldn't resolve the package 'rush_cli' in 'package:rush_cli/version.dart'.
Error: Couldn't resolve the package 'rush_prompt' in 'package:rush_prompt/rush_prompt.dart'.
bin/rush.dart:3:8: Error: Not found: 'package:args/command_runner.dart'
import 'package:args/command_runner.dart';
       ^
bin/rush.dart:4:8: Error: Not found: 'package:dart_console/dart_console.dart'
import 'package:dart_console/dart_console.dart';
       ^
bin/rush.dart:5:8: Error: Not found: 'package:rush_cli/commands/build_command/build_command.dart'
import 'package:rush_cli/commands/build_command/build_command.dart';
       ^
bin/rush.dart:6:8: Error: Not found: 'package:rush_cli/commands/create_command/create_command.dart'
import 'package:rush_cli/commands/create_command/create_command.dart';
       ^
bin/rush.dart:7:8: Error: Not found: 'package:rush_cli/commands/migrate_command/migrate_command.dart'
import 'package:rush_cli/commands/migrate_command/migrate_command.dart';
       ^
bin/rush.dart:8:8: Error: Not found: 'package:rush_cli/commands/upgrade_command/upgrade_command.dart'
import 'package:rush_cli/commands/upgrade_command/upgrade_command.dart';
       ^
bin/rush.dart:9:8: Error: Not found: 'package:rush_cli/helpers/app_data_dir.dart'
import 'package:rush_cli/helpers/app_data_dir.dart';
       ^
bin/rush.dart:10:8: Error: Not found: 'package:rush_cli/version.dart'
import 'package:rush_cli/version.dart';
       ^
bin/rush.dart:11:8: Error: Not found: 'package:rush_prompt/rush_prompt.dart'
import 'package:rush_prompt/rush_prompt.dart';
       ^
bin/rush.dart:55:33: Error: Type 'CommandRunner' not found.
class RushCommandRunner extends CommandRunner {
                                ^^^^^^^^^^^^^
bin/rush.dart:15:20: Error: Getter not found: 'RushDataDir'.
  final _dataDir = RushDataDir.dataDir()!;
                   ^^^^^^^^^^^
bin/rush.dart:28:18: Error: Method not found: 'CreateCommand'.
    ..addCommand(CreateCommand(_cd, _dataDir))
                 ^^^^^^^^^^^^^
bin/rush.dart:29:18: Error: Method not found: 'BuildCommand'.
    ..addCommand(BuildCommand(_cd, _dataDir))
                 ^^^^^^^^^^^^
bin/rush.dart:30:18: Error: Method not found: 'MigrateCommand'.
    ..addCommand(MigrateCommand(_cd, _dataDir))
                 ^^^^^^^^^^^^^^
bin/rush.dart:31:18: Error: Method not found: 'UpgradeCommand'.
    ..addCommand(UpgradeCommand(_dataDir))
                 ^^^^^^^^^^^^^^
bin/rush.dart:33:18: Error: 'UsageException' isn't a type.
      if (err is UsageException) {
                 ^^^^^^^^^^^^^^
bin/rush.dart:20:10: Error: The getter 'argParser' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'argParser'.
  runner.argParser.addFlag('version', abbr: 'v', negatable: false,
         ^^^^^^^^^
bin/rush.dart:28:7: Error: The method 'addCommand' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing method, or defining a method named 'addCommand'.
    ..addCommand(CreateCommand(_cd, _dataDir))
      ^^^^^^^^^^
bin/rush.dart:29:7: Error: The method 'addCommand' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing method, or defining a method named 'addCommand'.
    ..addCommand(BuildCommand(_cd, _dataDir))
      ^^^^^^^^^^
bin/rush.dart:30:7: Error: The method 'addCommand' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing method, or defining a method named 'addCommand'.
    ..addCommand(MigrateCommand(_cd, _dataDir))
      ^^^^^^^^^^
bin/rush.dart:31:7: Error: The method 'addCommand' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing method, or defining a method named 'addCommand'.
    ..addCommand(UpgradeCommand(_dataDir))
      ^^^^^^^^^^
bin/rush.dart:32:7: Error: The method 'run' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing method, or defining a method named 'run'.
    ..run(args).catchError((Object err) {
      ^^^
bin/rush.dart:42:3: Error: Method not found: 'PrintArt'.
  PrintArt();
  ^^^^^^^^
bin/rush.dart:43:3: Error: Method not found: 'Console'.
  Console()
  ^^^^^^^
bin/rush.dart:45:26: Error: Getter not found: 'ConsoleColor'.
    ..setForegroundColor(ConsoleColor.cyan)
                         ^^^^^^^^^^^^
bin/rush.dart:46:17: Error: Getter not found: 'rushVersion'.
    ..writeLine(rushVersion)
                ^^^^^^^^^^^
bin/rush.dart:49:26: Error: Getter not found: 'ConsoleColor'.
    ..setForegroundColor(ConsoleColor.cyan)
                         ^^^^^^^^^^^^
bin/rush.dart:50:17: Error: Getter not found: 'rushBuiltOn'.
    ..writeLine(rushBuiltOn)
                ^^^^^^^^^^^
bin/rush.dart:57:14: Error: Too many positional arguments: 0 allowed, but 2 found.
Try removing the extra positional arguments.
      : super(executableName, description);
             ^
bin/rush.dart:61:5: Error: The method 'PrintArt' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing method, or defining a method named 'PrintArt'.
    PrintArt();
    ^^^^^^^^
bin/rush.dart:63:21: Error: The method 'Console' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing method, or defining a method named 'Console'.
    final console = Console();
                    ^^^^^^^
bin/rush.dart:65:30: Error: The getter 'description' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'description'.
    console..writeLine(' ' + description)..writeLine();
                             ^^^^^^^^^^^
bin/rush.dart:70:28: Error: The getter 'ConsoleColor' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'ConsoleColor'.
      ..setForegroundColor(ConsoleColor.brightBlue)
                           ^^^^^^^^^^^^
bin/rush.dart:72:28: Error: The getter 'ConsoleColor' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'ConsoleColor'.
      ..setForegroundColor(ConsoleColor.cyan)
                           ^^^^^^^^^^^^
bin/rush.dart:74:28: Error: The getter 'ConsoleColor' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'ConsoleColor'.
      ..setForegroundColor(ConsoleColor.yellow)
                           ^^^^^^^^^^^^
bin/rush.dart:82:28: Error: The getter 'ConsoleColor' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'ConsoleColor'.
      ..setForegroundColor(ConsoleColor.yellow)
                           ^^^^^^^^^^^^
bin/rush.dart:92:28: Error: The getter 'ConsoleColor' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'ConsoleColor'.
      ..setForegroundColor(ConsoleColor.cyan)
                           ^^^^^^^^^^^^
bin/rush.dart:97:28: Error: The getter 'ConsoleColor' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'ConsoleColor'.
      ..setForegroundColor(ConsoleColor.cyan)
                           ^^^^^^^^^^^^
bin/rush.dart:102:28: Error: The getter 'ConsoleColor' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'ConsoleColor'.
      ..setForegroundColor(ConsoleColor.cyan)
                           ^^^^^^^^^^^^
bin/rush.dart:107:28: Error: The getter 'ConsoleColor' isn't defined for the class 'RushCommandRunner'.
 - 'RushCommandRunner' is from 'bin/rush.dart'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'ConsoleColor'.
      ..setForegroundColor(ConsoleColor.cyan)
                           ^^^^^^^^^^^^
Error: AOT compilation failed
Generating AOT kernel dill failed!

**Reply by shreyashsaitwal** [reply]
Ah, I forgot to mention that you need to run `dart pub get` before building. It should build successfully after that.

**Reply by HritikR** [reply]
Thanks alot @shreyashsaitwal. finally its working

**Reply by shreyashsaitwal** [reply]
@HritikR Sounds great! Can you send me the generated binary so I can do some tests with it?

**Reply by HritikR** [reply]
@shreyashsaitwal Sorry for replying late. Here's the link to [Download](https://drive.google.com/file/d/1Jts7u9K6XxUBju46Pyd-7C-lmbW3QuJ8/view?usp=sharing)

**Reply by shreyashsaitwal** [reply]
> Sorry for replying late.

No worries.
Can you please try if you can run [this ](https://raw.githubusercontent.com/shreyashsaitwal/pack/main/exe/mac/rush)executable? Don't add it to the `.rush/bin` directory. Just download it, `chmod +x` it, and try to run it.

**Reply by HritikR** [reply]
@shreyashsaitwal  it's working !!

**Reply by shreyashsaitwal** [reply]
Perfect! Now, can you replace the executable in `.rush/bin` directory with it and try to build some extension.

**Comment by shreyashsaitwal** [comment, marked as answer]
Fixed by [v1.2.1](https://github.com/shreyashsaitwal/rush-cli/releases/tag/v1.2.1)
